22.0.0.6-beta blog (#2247)
OpenLiberty/blogs: posts/2022-05-12-new-logging-functionality.adoc

blog_description: Open Liberty 22.0.0.6-beta offers one new beta feature (along with all GA features), which provides new logging functionality to join stack traces in logs, allowing for the stack trace emitted to be output as a single log event.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= New Logging functionality to join stack traces in logs in Open Liberty 22.0.0.6-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 22.0.0.6-beta offers one new beta feature (along with all GA features), which provides new logging functionality to join stack traces in logs, allowing for the stack trace emitted to be output as a single log event.

The link:{url-about}[Open Liberty] 22.0.0.6-beta includes the following beta features (along with all GA features):

* <<logging, New Logging functionality to join stack traces in logs>>

[#logging]
== New Logging functionality to join stack traces in logs

> The OpenLiberty logging when printing a stack trace to system error stream or system out stream it will print the individual lines of the stack trace as separate log events. This typically originates from `Throwable.printStackTrace()` calls from the application. This affects the console output, console logs , messages log and the trace logs. This new functionality introduced allows for the stack trace emitted to be output as a single log event. This is most impactful to users who forward their logs down stream to third party log analysis technologies. An example of this being the (E)lastic (L)ogstash (K)ibana stack. 

Reviewer: ```suggestion
Prior to this feature, when a stack trace was logged in Open Liberty, typically from `Throwable.printStackTrace()` calls from the application, each line of the stack trace was printed as a separate event.
The newly introduced functionality allows for the emitted stack trace to be output as a single log event. This is most impactful to users who forward their logs down stream to third party log analysis technologies. An example of this being the (E)lastic (L)ogstash (K)ibana stack.   This affects the console output, console logs , messages log and the trace logs.
```

*Examples:*

Console output *without* the stack traces joined:

> [source]

Reviewer: ```suggestion
[source, role="no_copy"]
```

----
[err] java.lang.ArithmeticException: divide by zero
[err]   at com.ibm.demo.MyResource.ex(MyResource.java:86)
[err]   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[err]   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[err]   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[err]   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[err]   at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.performInvocation(LibertyJaxRsServerFactoryBean.java:656)
[err]   at [internal classes]
[err]   at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
[err]   at com.ibm.websphere.jaxrs.server.IBMRestServlet.service(IBMRestServlet.java:96)
[err]   at [internal classes]
[err]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[err]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[err]   at java.base/java.lang.Thread.run(Thread.java:831)
----

Console output *with* the stack traces joined:

> [source]

Reviewer: ```suggestion
[source, role="no_copy"]
```

----
[err] java.lang.ArithmeticException: divide by zero
        at com.ibm.demo.MyResource.ex(MyResource.java:86)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.performInvocation(LibertyJaxRsServerFactoryBean.java:656)
        at [internal classes]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
        at com.ibm.websphere.jaxrs.server.IBMRestServlet.service(IBMRestServlet.java:96)
        at [internal classes]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:831)
----

`messages.log` output *without* the stack traces joined:

> [source]

Reviewer: ```suggestion
[source, role="no_copy"]
```

[2022-04-06, 15:50:22:248 EDT] 00000047 SystemErr                                                    R 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2022-04-06, 15:50:22:249 EDT] 00000047 SystemErr                                                    R 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2022-04-06, 15:50:22:250 EDT] 00000047 SystemErr                                                    R 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2022-04-06, 15:50:22:251 EDT] 00000047 SystemErr                                                    R 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[2022-04-06, 15:50:22:251 EDT] 00000047 SystemErr                                                    R 	at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.performInvocation(LibertyJaxRsServerFactoryBean.java:656)
...
----

`messages.log` output *with* the stack traces joined:

> [source]

Reviewer: ```suggestion
[source, role="no_copy"]
```

----
[2022-04-06, 15:52:38:586 EDT] 00000077 SystemErr                                                    R java.lang.ArithmeticException: divide by zero
	at com.ibm.demo.MyResource.ex(MyResource.java:86)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.performInvocation(LibertyJaxRsServerFactoryBean.java:656)
        ...
----

`messages.log` output *without* the stack traces joined using `JSON` logging:

> [source, json]

Reviewer: ```suggestion
[source, json, role="no_copy"]
```

{"type":"liberty_message","host":"LAPTOP-JU4FJ7TJ","ibm_userDir":"C:\/devdir\/LibertiesFeb18\/open-liberty\/dev\/build.image\/wlp\/usr\/","ibm_serverName":"sj","message":"\tat com.ibm.demo.MyResource.ex(MyResource.java:86)","ibm_threadId":"00000034","ibm_datetime":"2022-04-20T13:41:37.616-0400","module":"SystemErr","loglevel":"SystemErr","ibm_methodName":"","ibm_className":"","ibm_sequence":"1650476497616_000000000006A","ext_thread":"Default Executor-thread-2"}
{"type":"liberty_message","host":"LAPTOP-JU4FJ7TJ","ibm_userDir":"C:\/devdir\/LibertiesFeb18\/open-liberty\/dev\/build.image\/wlp\/usr\/","ibm_serverName":"sj","message":"\tat java.base\/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)","ibm_threadId":"00000034","ibm_datetime":"2022-04-20T13:41:37.626-0400","module":"SystemErr","loglevel":"SystemErr","ibm_methodName":"","ibm_className":"","ibm_sequence":"1650476497626_000000000006B","ext_thread":"Default Executor-thread-2"}
...
----

`messages.log` output *with* the stack traces joined using `JSON` logging:

> [source, json]

Reviewer: ```suggestion
[source, json, role="no_copy"]
```

----

This new functionality is enabled by configuration through either the a bootstrap property, environment variable and/or through the `server.xml`. Note that if there is configuration present in all strategies mentioned then the configuration precedence is the `server.xml` > environment variable > bootstrap property.

> *Configuration:*
> 
> `Bootstrap.properties` > `com.ibm.ws.logging.stackJoin=true/false`
> `Server.env` > `WLP_LOGGING_STACK_JOIN=true/false`
> `Server.xml` > `<logging stackJoin="true/false" />`

Reviewer: ```suggestion
*Configuration properties:*

- `bootstrap.properties`: 
[source]
----
`com.ibm.ws.logging.stackJoin=true
----
- `server.env`: 
[source]
----
WLP_LOGGING_STACK_JOIN=true/false
----
- `server.xml`: 
[source,xml]
----
<logging stackJoin="true/false" />
----
```

To enable the new beta features in your app, add them to your `server.xml`:

[#run]
=== Try it now 

> To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE XX, Java SE 17, Java SE 11, and Java SE 8.

Reviewer: ```suggestion
To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 18, Java SE 17, Java SE 11, and Java SE 8.
```
